### ISSUE TITLE
Meta: active memory limiting

### ISSUE THREAD

**Issue by emorice** [emorice/galp#28]
Status: divided in sub-issues:
 * [x] #29
 * [ ] to open: user-mode oom killer
 * [ ] to open: memory-based message withholding

Solutions to #8 like #10 and #26 are reaction-based: they kick in after oom kills, but they don't actually limit memory in any way. Efforts like #25 help with the root cause, but in the end how much memory a task uses is outside the scope of galp and requires case-by-case optimization, there is no generic way to make sure something runs with a given amount of memory -- well, apart from swapping, but we cannot do that ourselves.

Reaction methods cannot protect galp from the massive, instantaneous, general SIGBUSes that we observe when we exhaust the cgroup memory despite the efforts of the OOM killer ; and we cannot make the tasks just use less memory, so we need something in between: prevent tasks from running before we use up too much memory.

Inside this group of solutions, we can refine:
 * Soft solutions: we try to schedule things better
 * Hard solutions: we abort tasks if needed

For now, I care mostly about hard solutions, because ultimately scheduling better requires knowing how much memory you're going to use, and you can really know this only after you actually try to run it. In slurm/lsf/sge, you can schedule things to run smoothly with memory allocations once you've set them, and you know what to set them to because it crashed with less memory allocated the time before that. Here it's the same, scheduling well will save you if you know how much you need, and to get there you're going to under-allocate first, so you need to deal with bad scheduling anyway -- since failing here means crashing the whole galp, and you don't want to crash galp once per task you need to set memory for...

There are three places where we can act:
  * Submissions. Even when we don't know how much memory things will need, we can still block them based on current usage. This means monitoring memory use, and withholding new tasks if it's too high. This will not save us in all cases, because the memory usage will often go up some time after the task is started, and we may already have launched too many tasks at that point.
  * External aborts. We can send messages, interrupts of even kill processes if too much memory is used. This has a shorter reaction time that scheduling, but the idea is the same: we monitor memory use, and cancel tasks in one way or an other if it's too high. Essentially, it's a usermode oom kill.
  * Internal aborts. I had forgotten about it, but I just tested it quickly with numpy and it seems to actually work ! Instead of reacting to tasks allocating too much memory, there's a system feature we can use to stop them the instant they try to use more than we want: `setrlimit`, available in python through the `resource` module. With this, we can set a limit on a process, and this will cause `MemoryError`s exception to be raised inside the client code. This is by far the safest of all three options, because if all processes at any given time have a total limit set to something acceptable, system oom kills becomes much less likely and complete galp crashes should be mostly preventable.

As far as I know `setrlimit` only works with virtual memory, but empirically on our pipeline steps the rss seems to track the vms rather closely (often more than 90% of the vm is actually used) so it's still very useful. That being said, this suggests that the ultimate solution would be a combination of all:
  * Internal limit is the only thing that can really prevent system crashes in the end.
  * Usermode oom-like things is the only way to limit in real time things that cannot be limited up-front, like rss or non-peak memory size (ah, yes, with setrlimit-based solutions, you would set the limit typically at the beginning of the task, so if your task runs most of the time at a much lower average use you could end up wasting a lot of memory)
  * Withholding tasks is the only thing that can keep your memory use low enough to give the other tasks a chance to finish, the two other involve killing running tasks but you would end up filling your system with tasks that would get killed anyway.

In conclusion, among these the `setrlimit` solution takes priority since it's the shortest path to a non-crashing galp system.
